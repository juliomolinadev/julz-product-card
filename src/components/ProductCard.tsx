import React, { createContext } from 'react';

import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import {
  InitialValues,
  Product,
  ProductCardHandler,
  ProductContextProps,
  onChangeArgs,
} from '../interfaces/interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  // children?: React.ReactElement | React.ReactElement[];
  product: Product;
  children: (args: ProductCardHandler) => JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,

          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
