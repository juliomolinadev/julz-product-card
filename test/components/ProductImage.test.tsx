import React from 'react';
import render from 'react-test-renderer';

import { ProductImage, ProductCard } from '../../src/components';
import { product2 } from '../data/products';

describe('<ProductImage /> tests', () => {
  // test('should show component with custom image', () => {
  //   const wrapper = render.create(<ProductImage img={'https://hello.jpg'} />);

  //   expect(wrapper.toJSON()).toMatchSnapshot();
  // });

  test('should show component with product image', () => {
    const wrapper = render.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
