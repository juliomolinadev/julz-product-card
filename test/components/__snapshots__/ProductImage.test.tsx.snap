// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductImage /> tests should show component with product image 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffee mug"
    className="productImg undefined"
    src="./coffe-mug.jpg"
  />
</div>
`;
